File: web/src/core/descriptor.ts
// Copyright (C) 2026 Derek Quenneville
// SPDX-License-Identifier: GPL-2.0-or-later

import {
  DESCRIPTOR_MAGIC,
  DESCRIPTOR_OFFSET,
  DESCRIPTOR_SIZE,
  DESCRIPTOR_VERSION,
  ROM_BASE,
} from "./constants.ts";

/** The template's 32-byte descriptor at ROM offset 0x20 (cart/README.md). */
export interface TemplateDescriptor {
  descriptorVersion: number;
  /** The pack format version the cartridge program reads. */
  packVersion: number;
  /** Address the pack must start at. */
  packBase: number;
  /** End of the cartridge address window (4 MB for the Floopy Drive). */
  romLimit: number;
  buildId: string;
}

export class TemplateError extends Error {
  override name = "TemplateError";
}

function ascii(bytes: Uint8Array): string {
  return String.fromCharCode(...bytes);
}

export function readDescriptor(template: Uint8Array): TemplateDescriptor {
  if (template.byteLength < DESCRIPTOR_OFFSET + DESCRIPTOR_SIZE) {
    throw new TemplateError("The template is too short to be a Loopy Photo Cart ROM.");
  }
  const view = new DataView(template.buffer, template.byteOffset, template.byteLength);

  const magic = ascii(template.subarray(DESCRIPTOR_OFFSET, DESCRIPTOR_OFFSET + 4));
  if (magic !== DESCRIPTOR_MAGIC) {
    throw new TemplateError("This is not a Loopy Photo Cart template (no LPCT descriptor).");
  }

  const descriptorVersion = view.getUint16(DESCRIPTOR_OFFSET + 4);
  if (descriptorVersion !== DESCRIPTOR_VERSION) {
    throw new TemplateError(`Unsupported template descriptor version ${descriptorVersion}.`);
  }

  const packVersion = view.getUint16(DESCRIPTOR_OFFSET + 6);
  const packBase = view.getUint32(DESCRIPTOR_OFFSET + 8);
  const romLimit = view.getUint32(DESCRIPTOR_OFFSET + 12);
  if (packBase <= ROM_BASE || packBase % 4 !== 0 || romLimit <= packBase) {
    throw new TemplateError("The template descriptor's pack region is invalid.");
  }

  const idBytes = template.subarray(DESCRIPTOR_OFFSET + 16, DESCRIPTOR_OFFSET + 32);
  const nul = idBytes.indexOf(0);
  const buildId = ascii(nul < 0 ? idBytes : idBytes.subarray(0, nul));

  return { descriptorVersion, packVersion, packBase, romLimit, buildId };
}

File: cart/src/descriptor.c
//
// Copyright(C) 2026 Derek Quenneville
//
// You can redistribute and/or modify this program under the terms of the
// GNU General Public License version 2 as published by the Free Software
// Foundation, or any later version. This program is distributed WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
//
// The template descriptor: 32 bytes at ROM offset 0x20, in the free space the
// cartridge header leaves before the vector table (tools/loopy.ld).
//
// The web app reads this before it appends a photo pack, and refuses a
// template whose pack format it does not write. That is what stops a stale
// cached template from silently producing a ROM that cannot read its photos.
//
// Layout, big-endian (the SH-1's native order):
//   0x20  char[4]  "LPCT"
//   0x24  u16      descriptor version
//   0x26  u16      pack format version the program reads
//   0x28  u32      pack base address
//   0x2C  u32      cartridge limit (end of the 4 MB window)
//   0x30  char[16] build id, NUL-padded
//

#include "lpc.h"

struct lpc_descriptor {
	char     magic[4];
	uint16_t desc_version;
	uint16_t pack_version;
	uint32_t pack_base;
	uint32_t rom_limit;
	char     build_id[16];
};

_Static_assert(sizeof(struct lpc_descriptor) == 32, "descriptor is 32 bytes");

__attribute__((section(".header.descriptor"), used))
const struct lpc_descriptor lpc_descriptor = {
	{ 'L', 'P', 'C', 'T' },
	LPC_DESC_VERSION,
	LPC_PACK_VERSION,
	LPC_PACK_BASE,
	LPC_ROM_LIMIT,
	LPC_BUILD_ID,
};
